### PR TITLE
New metric: Cohen's kappa

### LINKED ISSUE #15285: Add kappa coefficient as a new metric?
Currently tensorflow supports using accuracy as a metric for model's performance.

However, for unbalanced datasets, kappa coefficient (https://www.wikiwand.com/en/Cohen%27s_kappa#) is a commonly used metric. Would it be possible to add this one in the model.metrics?

### PR DESCRIPTION
**Pull request by facaiy**
resolve #15285.

Add new metric: `cohen_kappa`, which is equivalent to `sklearn.metrics.cohen_kappa_score`(>=0.19), but the implementation doesn't support weighted matrix yet.

Ref:
+ [Cohen's kappa - Wiki](https://en.wikipedia.org/wiki/Cohen's_kappa)
+ [Cohen's kappa: Index of Inter-rater Reliability](http://psych.unl.edu/psycrs/handcomp/hckappa.PDF)
+ [sklearn.metrics.cohen_kappa_score](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.cohen_kappa_score.html)

### How to test

+ [x] add test cases.
+ [ ] pass all tests.

### CODE REVIEW

**Comment by facaiy** [comment]
I checked the log, and found that scikit-learn (>= 0.19) has not been deploy in CI environments.
```bash
tensorflow/contrib/metrics/python/ops/metric_ops_test.py:6799: [E1123(unexpected-keyword-arg), CohenKappaTest.testWeighted] Unexpected keyword argument 'sample_weight' in function call
```

Because `sklearn.metrics.cohen_kappa_score` has supported `sample_weight` from 0.19 (August 12, 2017), and the function is necessary to check `weights` of `cohen_kappa` we proposed. So could we upgrade the scikit-learn dependency?

File: tensorflow/contrib/metrics/python/ops/metric_ops_test.py [hunk 1]
```diff
@@ -6660,5 +6661,194 @@ def test2dWeightedValues_placeholders(self):
...
+                np.reshape(predictions, shape), dtype=dtype)
+            labels_tensor = constant_op.constant(
+                np.reshape(labels, shape), dtype=dtype)
+            kappa, update_op = metrics.cohen_kappa(
+                labels_tensor, predictions_tensor, 3, weights=weight)
+
+            sess.run(variables.local_variables_initializer())
+            self.assertAlmostEqual(0.45, sess.run(update_op), 2)
+            self.assertAlmostEqual(0.45, kappa.eval(), 2)
+
```
Comment (roumposg) [inline tensorflow/contrib/metrics/python/ops/metric_ops_test.py]: Can you add a short comment about how 0.45 is calculated? Thanks.

File: tensorflow/contrib/metrics/python/ops/metric_ops_test.py [hunk 2]
```diff
@@ -6660,5 +6661,194 @@ def test2dWeightedValues_placeholders(self):
...
+  def testBasic2(self):
+    labels = np.random.randint(0, 4, size=(100, 1))
+    predictions = np.random.randint(0, 4, size=(100, 1))
+    expect = sk_metrics.cohen_kappa_score(labels, predictions)
```
Comment (roumposg) [inline tensorflow/contrib/metrics/python/ops/metric_ops_test.py]: We try to avoid random in tests. I think you can just skip this test, since testBasic tests the base test case.

File: tensorflow/contrib/metrics/python/ops/metric_ops_test.py [hunk 3]
```diff
@@ -6660,5 +6661,194 @@ def test2dWeightedValues_placeholders(self):
...
+      labels = constant_op.constant(labels)
+      kappa, update_op = metrics.cohen_kappa(labels, predictions, 4)
+
+      sess.run(variables.local_variables_initializer())
+      self.assertAlmostEqual(expect, sess.run(update_op), 5)
+      self.assertAlmostEqual(expect, kappa.eval(), 5)
+
+  def testAllCorrect(self):
+    inputs = np.random.randint(0, 4, size=(100, 1))
+    expect = sk_metrics.cohen_kappa_score(inputs, inputs)
```
Comment (roumposg) [inline tensorflow/contrib/metrics/python/ops/metric_ops_test.py]: Similarly, we want to avoid random. You can just use a small matrix for labels and predictions.

File: tensorflow/contrib/metrics/python/ops/metric_ops_test.py [hunk 4]
```diff
@@ -6660,5 +6661,194 @@ def test2dWeightedValues_placeholders(self):
...
+
+      sess.run(variables.local_variables_initializer())
+      self.assertAlmostEqual(expect, sess.run(update_op), 5)
+      self.assertAlmostEqual(expect, kappa.eval(), 5)
+
+  def testWeighted(self):
+    labels = np.random.randint(0, 4, size=(100, 1))
+    predictions = np.random.randint(0, 4, size=(100, 1))
+    weights = np.random.random(size=(100, 1))
+    expect = sk_metrics.cohen_kappa_score(labels, predictions,
```
Comment (roumposg) [inline tensorflow/contrib/metrics/python/ops/metric_ops_test.py]: ditto

File: tensorflow/contrib/metrics/python/ops/metric_ops_test.py [hunk 5]
```diff
@@ -6660,5 +6661,194 @@ def test2dWeightedValues_placeholders(self):
...
+      predictions = constant_op.constant(predictions, dtype=dtypes_lib.float32)
+      labels = constant_op.constant(labels)
+      kappa, update_op = metrics.cohen_kappa(labels, predictions, 4,
+                                             weights=weights)
+
+      sess.run(variables.local_variables_initializer())
+      self.assertAlmostEqual(expect, sess.run(update_op), 5)
+      self.assertAlmostEqual(expect, kappa.eval(), 5)
+
+  def testWithMultipleUpdates(self):
```
Comment (roumposg) [inline tensorflow/contrib/metrics/python/ops/metric_ops_test.py]: Can you wrap the lines in this test to 80 chars? Thanks.

File: tensorflow/contrib/metrics/python/ops/metric_ops.py [hunk 6]
```diff
@@ -3297,9 +3299,120 @@ def count(values,
     return count_, update_op
 
 
+def cohen_kappa(labels, predictions, num_classes, weights=None,
```
Comment (roumposg) [inline tensorflow/contrib/metrics/python/ops/metric_ops.py]: Can you make this argument optional, similar to the confusion_matrix method? Thanks.

File: tensorflow/contrib/metrics/python/ops/metric_ops_test.py (hunk 2 above)
Comment (roumposg) [inline tensorflow/contrib/metrics/python/ops/metric_ops_test.py]: I would prefer if we don't depend on sk_metrics, since the calculation is relatively simple if the data are small. This would also solve the problem you found importing sklearn.

**Review by roumposg — COMMENTED** [review]
Thanks for working on that! I have a few comments, mostly straightforward.

File: tensorflow/python/ops/metrics_impl.py [hunk 7]
```diff
@@ -185,11 +185,11 @@ def _safe_div(numerator, denominator, name):
...
-  return array_ops.where(
-      math_ops.greater(denominator, 0),
-      math_ops.truediv(numerator, denominator),
-      0,
-      name=name)
+  t = math_ops.truediv(numerator, denominator)
+  zero = array_ops.zeros_like(t, dtype=denominator.dtype)
+  contion = math_ops.greater(denominator, zero)
+  zero = math_ops.cast(zero, t.dtype)
+  return array_ops.where(contion, t, zero, name=name)
```
[the remark this answers is not in the document]
Comment (facaiy) [reply tensorflow/python/ops/metrics_impl.py]: The old version requires that `numerator` must be a scalar, and is incompatible with float64. The change makes it more general.

File: tensorflow/contrib/metrics/python/ops/metric_ops_test.py (hunk 2 above)
Comment (facaiy) [reply tensorflow/contrib/metrics/python/ops/metric_ops_test.py]: I think sk_metrics give us more secure, however it is OK to calculate the `expect` by ourselves.

File: tensorflow/contrib/metrics/python/ops/metric_ops.py [hunk 8]
```diff
@@ -3297,9 +3299,120 @@ def count(values,
...
+    # Table of the counts of agreement:
+    counts_in_table = confusion_matrix.confusion_matrix(
+      labels, predictions,
+      num_classes=num_classes, weights=weights,
+      dtype=stat_dtype, name="counts_in_table")
+
+    po = array_ops.diag_part(counts_in_table, name='po')
+    pe_row = math_ops.reduce_sum(counts_in_table, axis=0, name='pe_row')
+    pe_col = math_ops.reduce_sum(counts_in_table, axis=1, name='pe_col')
+    with ops.control_dependencies([po, pe_row, pe_col]):
```
[the remark this answers is not in the document]
Comment (facaiy) [reply tensorflow/contrib/metrics/python/ops/metric_ops.py]: Yes, I agree with you.

File: tensorflow/contrib/metrics/python/ops/metric_ops.py [hunk 9]
```diff
@@ -3297,9 +3299,120 @@ def count(values,
...
+    raise RuntimeError('tf.contrib.metrics.cohen_kappa is not supported'
+                       'when eager execution is enabled.')
+  if num_classes < 2:
+    raise ValueError('`num_classes` must be greater than 1.'
+                     'Found: {}'.format(num_classes))
+  with variable_scope.variable_scope(name, 'cohen_kappa',
+                                     (labels, predictions, weights)):
+    # Convert 2-dim (num, 1) to 1-dim (num,)
+    labels.get_shape().with_rank_at_most(2)
+    labels = array_ops.squeeze(labels)
```
[the remark this answers is not in the document]
Comment (facaiy) [reply tensorflow/contrib/metrics/python/ops/metric_ops.py]: Sadly, `confuse_matrix` only supports 1-D tensor. If we don't squeeze labels explicitly, I'm afraid we cannot support 2-dim like `(num, 1)`.

File: tensorflow/contrib/metrics/python/ops/metric_ops.py (hunk 6 above)
Comment (facaiy) [reply tensorflow/contrib/metrics/python/ops/metric_ops.py]: Because tensorflow supports batch data, the maximum label might vary with each batch, hence I think it might be unsafe to let `num_classes` optional.

**Comment by facaiy** [comment]
@roumposg Thanks for your review and comment. I have revised those codes, except of two unresolved comments.

File: tensorflow/contrib/metrics/python/ops/metric_ops.py [hunk 10]
```diff
@@ -3326,8 +3329,10 @@ def cohen_kappa(labels, predictions, num_classes, weights=None,
   doesn't support weighted matrix yet.
 
   Args:
-    labels: 1-D `Tensor` of the ground truth values.
-    predictions: 1-D `Tensor` of the predicted values.
+    labels: 1-D `Tensor` of real labels for the classification task. Must be
+      one of the following types: int16, int32, int64, float32, float64.
```
Comment (roumposg) [inline tensorflow/contrib/metrics/python/ops/metric_ops.py]: Can we drop float? I thought these are class indices. Also, for integer predictions, we prefer the arg name `predictions_idx`, as in precition_at_top_k.

File: tensorflow/contrib/metrics/python/ops/metric_ops.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (roumposg) [reply tensorflow/contrib/metrics/python/ops/metric_ops.py]: I thought that _remove_squeezable_dimensions below does the trick, but I see you are right.

How about changing to this statement to make it safer:
labels = array_ops.squeeze(labels, axis=[-1])

File: tensorflow/python/ops/metrics_impl.py [hunk 11]
```diff
@@ -187,9 +187,9 @@ def _safe_div(numerator, denominator, name):
   """
```
Comment (roumposg) [inline tensorflow/python/ops/metrics_impl.py]: Can you update the method description to "Divides two tensors element-wise, returning 0 ..."?

**Review by roumposg — CHANGES_REQUESTED** [review]
Thanks! A few remaining nits only.

**Comment by facaiy** [comment]
Thank you, @roumposg. I think all comments left have been resolved. Could you take a look again?

**Review by roumposg — APPROVED** [review]

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
@caisq any idea what the `math_grad_test` problem is? It seems to happen on multiple PRs.

**Comment by drpngx** [comment]
Maybe it's some flakiness. Will take a look later unless you know what it is.

**Comment by caisq** [comment]
@drpngx : that looks like a flaky test. we should keep an eye on it.

**Review by drpngx — APPROVED** [review]
As per @roumposg (googler).
